libnetwork/d/overlay: fix logical race conditions

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

- For #49908 
- Split from #50081 
- Stacked on top of #50106 

**- What I did**
**- How I did it**

The concurrency control in the overlay driver is logically unsound. While the use of mutexes is sufficient to prevent data races -- violations of the Go memory model -- many operations which need to be atomic are performed with unbounded concurrency. For instance, programming the encryption parameters for a peer can race with encryption keys being updated, which could lead to inconsistencies between the parameters programmed into the kernel and the desired state.

Overhaul the use of locks in the overlay network driver. Implement sound locking at the network granularity: operations may proceed concurrently iff they are being applied to distinct networks. Push the responsibility of locking up to the code which calls methods or accesses struct fields to avoid deadlock situations like we had previously with d.initSandboxPeerDB() and to make the code easier to reason about.

Each overlay network has a distinct peer db. The NetworkDB watch for the overlay peer table for the network will only start after (*driver).CreateNetwork returns and will be stopped before libnetwork calls (*driver).DeleteNetwork, therefore the lifetime of the peer db for a network is constrained to the lifetime of the network itself. Yet the peer db for a network is tracked in a dedicated map, separately from the network objects themselves. This has resulted in a parallel set of mutexes to manage concurrency of the peer db distinct from the mutexes for the driver and networks. Move the peer db for a network into a field of the network struct and guard it from concurrent access using the per-network lock. Move the methods for manipulating the peer db into the network struct so that the methods can only be called if the caller has a reference to the network object.

Network creation and deletion is synchronized using the driver-scope mutex, but some of the kernel programming is performed outside of the critical section. It is possible for network deletion to race with recreating the network, interleaving the kernel programming for the network creation and deletion, resulting in inconsistent kernel state. Parallelize network creation and deletion soundly. Use a double-checked locking scheme to soundly handle the case of concurrent CreateNetwork and DeleteNetwork for the same network id without blocking operations on other networks. Synchronize operations on a network so that operations on the network such as adding a neighbor to the peer db are performed atomically, not interleaved with deleting the network.

Handle encryption-key updates in a critical section so they can no longer be interleaved with kernel programming of encryption parameters.

**- How to verify it**
1. Create a multi-node Swarm cluster.
2. Create a user-defined overlay network.
3. `docker service create --mode global --network my-overlay -p 80:8080 nginxdemos/hello:plain-text`
4. Exec into one of the service containers and verify that the other containers can be curl'ed by the IP address of its user-defined overlay network endpoint. Repeat with the other replicas.
5. curl port 8080 on each Swarm node. Verify that the service is reachable. Repeat to verify that the requests are load-balanced among the replicas.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
- Improve the reliability of the overlay network driver
```

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libnetwork/d/overlay: un-embed mutexes

It is easier to find all references when they are struct fields rather
than embedded structs.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork/d/overlay: fix logical race conditions

The concurrency control in the overlay driver is logically unsound.
While the use of mutexes is sufficient to prevent data races --
violations of the Go memory model -- many operations which need to be
atomic are performed with unbounded concurrency.

Overhaul the use of locks in the overlay network driver. Implement sound
locking at the network granularity: operations may proceed concurrently
iff they are being applied to distinct networks. Push the responsibility
of locking up to the code which calls methods or accesses struct fields
to avoid deadlock situations like we had previously with
d.initSandboxPeerDB() and to make the code easier to reason about.

Each overlay network has a distinct peer db. The NetworkDB watch for the
overlay peer table for the network will only start after
(*driver).CreateNetwork returns and will be stopped before libnetwork
calls (*driver).DeleteNetwork, therefore the lifetime of the peer db for
a network is constrained to the lifetime of the network itself. Yet the
peer db for a network is tracked in a dedicated map, separately from the
network objects themselves. This has resulted in a parallel set of
mutexes to manage concurrency of the peer db distinct from the mutexes
for the driver and networks. Move the peer db for a network into a field
of the network struct and guard it from concurrent access using the
per-network lock. Move the methods for manipulating the peer db into the
network struct so that the methods can only be called if the caller has
a reference to the network object.

Network creation and deletion are synchronized using the driver-scope
mutex, but some of the kernel programming is performed outside of the
critical section. It is possible for network deletion to race with
recreating the network, interleaving the kernel programming for the
network creation and deletion, resulting in inconsistent kernel state.
Parallelize network creation and deletion soundly. Use a double-checked
locking scheme to soundly handle the case of concurrent CreateNetwork
and DeleteNetwork for the same network id without blocking operations
on other networks. Synchronize operations on a network so that
operations on the network such as adding a neighbor to the peer db are
performed atomically, not interleaved with deleting the network.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on libnetwork/drivers/overlay/overlay.go:
```diff
@@ -29,27 +29,30 @@ const (
+	config map[string]interface{}
+	initOS sync.Once
+
+	// Has dedicated locks; use the locks inside the structs when accessing them.
+
+	secMap encrMap
```
I guess this now applies to `secMap` (not the other things further down), so we may as well make it a "godoc" for that field;

```suggestion
	// secMap has dedicated locks; use the locks inside the structs when accessing them.
	secMap encrMap
```

corhere replied:
The comment applies to all fields under the section-heading comment. There just happens to only be one field in the section at the moment. A new section heading comment "clears" the preceding comments. Nobody is going to be confused about the immutability of secMap just because there is a comment about immutability some lines above.

thaJeztah reviewed (commented):
Looks good! I left some thoughts / ideas; happy to hear your thoughts!

corhere commented on libnetwork/drivers/overlay/ov_network.go:
```diff
@@ -596,18 +613,34 @@ func (n *network) initSandbox() error {
-	return n.sbox
+	for {
+		if n == nil {
+			return nil
```
That isn't friendly to logging a structured error

corhere commented on libnetwork/drivers/overlay/joinleave.go:
```diff
@@ -35,18 +35,24 @@ func (d *driver) Join(ctx context.Context, nid, eid string, sboxKey string, jinf
-	if n.secure && len(d.keys) == 0 {
-		return errors.New("cannot join secure network: encryption keys not present")
+	if n.secure {
+		d.secMap.mu.Lock()
```
secMap doesn't have any methods (aside from implementing the Stringer interface) -- its internals and its interface are one and the same! There is no avoiding "diving into its internals" so I see little benefit to adding boilerplate around locking a mutex.

thaJeztah reviewed (approved):
LGTM


Left a comment about the `walkNodes` method; non-blocking, but let me know if they make sense.

corhere commented on libnetwork/drivers/overlay/encryption.go:
```diff
@@ -451,39 +456,41 @@ func buildAeadAlgo(k *key, s int) *netlink.XfrmStateAlgo {
-func (d *driver) secMapWalk(f func(netip.Addr, []spi) ([]spi, bool)) error {
-	d.secMap.Lock()
-	for rIP, node := range d.secMap.nodes {
+func (e *encrMap) WalkNodes(f func(netip.Addr, []spi) ([]spi, bool)) error {
```
It's only used internally, there's only one call site, the error return isn't used, the `stop bool` feature isn't used, and I find it surprising that a function named walk mutates while iterating. I'm going to inline the loop into the caller instead.

thaJeztah replied:
Yup; inlining would've been my alternative - I also noticed it only had a single use. This may have been a remnant of some of the "walker" interfaces that libnetwork used to have defined (I recall a bunch of things in https://github.com/moby/moby/pull/46206), or just generic libnetwork randomness 😂

corhere committed:
libn/d/overlay: inline secMapWalk into only caller

func (*driver) secMapWalk is a curious beast. It is named walk, yet it
also mutates the collection being iterated over. It returns an error,
but that error is always nil. It takes a callback that can break
iteration, yet the only caller makes no use of that affordance. Its
utility is limited and the abstraction hinders readability more than it
helps. Open-code the d.secMap.nodes loop into
func (*driver) updateKeys(), the only caller.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah reviewed (approved):
still LGTM, thx!

@robmry @akerouanton ptal

robmry reviewed (approved):
LGTM - just nits/questions ...

corhere committed:
libn/d/overlay: fix encryption race conditions

There is a dedicated mutex for synchronizing access to the encrMap.
Separately, the main driver mutex is used for synchronizing access to
the encryption keys. Their use is sufficient to prevent data races (if
used correctly, which is not the case) but not logical race conditions.
Programming the encryption parameters for a peer can race with
encryption keys being updated, which could lead to inconsistencies
between the parameters programmed into the kernel and the desired state.

Introduce a new mutex for synchronizing encryption operations. Use that
mutex to synchronize access to both encrMap and keys. Handle encryption
key updates in a critical section so they can no longer be interleaved
with kernel programming of encryption parameters.

Signed-off-by: Cory Snider <csnider@mirantis.com>

robmry reviewed (approved):
Thank you - still LGTM.

akerouanton commented:
I have a few pending comments but I need to take another look tomorrow.

akerouanton reviewed (approved):
Beside my two comments, LGTM.

corhere commented on libnetwork/drivers/overlay/joinleave.go:
```diff
@@ -193,12 +199,30 @@ func (d *driver) EventNotify(etype driverapi.EventType, nid, tableName, key stri
+	if etype == driverapi.Delete {
+		opname = "delete"
+		err = n.peerDelete(eid, addr, mac, vtep)
+	} else {
```
That's the way it was before I started refactoring. 🤷 
I only touched this part of the code to refactor the error handling. The control flow is the same as it was before.
https://github.com/moby/moby/blob/aaf3e76ef2bd37cf2e57e9ad9b72cb632d28dd99/libnetwork/drivers/overlay/joinleave.go#L196-L201

Come to think of it, it is likely subtly incorrect to handle creates the same as updates. While the driver never performs an update operation on an `overlay_peer_table` entry (libnetwork doesn't afford a mechanism for drivers to do so) it could still be problematic if it is possible for the same endpoint ID to leave and rejoin a network, as the DELETE and re-CREATE could be coalesced into a single UPDATE if a peer receives the CREATE event first. This might be one of the root causes of #49908! I should fix that.
